HelpersTask676_Add_script_to_synchronize_GH_branch_protection_and_merge_property

sandeepthalapanane opened the pull request:
Addressing #676 

Documentation - https://docs.google.com/document/d/1Kpe-DVpBjcy7Do4-NmnxlJkRcwjSMqwT5GFpjSE2qvo/edit?usp=sharing

I have included a way to revert GitHub settings to their defaults if they are present in the repo but not listed in the YAML manifest using the `switch_default` flag.

If a field is not present in the YAML file, `NotSet` is used to ensure no accidental overwrites.

sandeepthalapanane commented:
@heanhsok, could you please review this code?

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+        branches = repo.get_branches()
+        for branch in branches:
+            try:
+                # Get branch protection settings.
+                protection = branch.get_protection()
+                # Remove branch protection if branch exists in repo but not in settings.
+                if branch.name not in self.branch_protection:
+                    if execute:
+                        protection.remove()
+                        _LOG.info(
+                            "Branch protection removed from %s", branch.name
+                        )
+                    else:
+                        _LOG.info(
+                            "Would remove branch protection from %s without --dry_run",
+                            branch.name,
+                        )
```
This will need a bit of explanation. For example, why do we remove branch protection if branch exists in repo but not in setting?

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+        current_settings = self.get_repository_settings(repo)
+
+        # Check each setting in the repository.
+        for setting, value in current_settings.items():
+            if setting not in self.repo_settings and value:
+                if execute:
+                    if setting in [
+                        "allow_merge_commit",
+                        "allow_squash_merge",
+                        "allow_rebase_merge",
+                        "delete_branch_on_merge",
+                        "has_issues",
+                        "has_wiki",
+                        "has_projects",
+                    ]:
+                        # Reset setting to default value.
+                        repo.edit(**{setting: False})
+                        _LOG.info(
+                            "Reset repository setting %s to default", setting
+                        )
```
Can we add some explanation, too? For example, why are only certain parameters chosen?

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+                        "teams", gh_obj.NotSet
+                    ),
+                )
+                _LOG.info("Branch protection rules applied to %s", branch_name)
```
Can we show what was changed as well?

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+                )
+                _LOG.info("Branch protection rules applied to %s", branch_name)
+            else:
+                _LOG.info(
+                    "Would apply branch protection rules to %s without --dry_run",
+                    branch_name,
+                )
```
To make this more useful, we want to show exactly what will be changed so the user can see the diff

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+        Apply repository settings.
+
+        :param repo: github repository object
+        :param execute: whether to do dry run
```
Why not just call it `dry_run` so it's the same as the parameter that was passed by the user?

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+                repo.replace_topics(topics)
+                _LOG.info("Updated repository topics")
+
+            _LOG.info("Repository settings applied")
+        else:
+            _LOG.info("Would apply repository settings without --dry_run")
```
Same one. We want to see what exactly will be applied if it's a dry run, and what has been applied otherwise.

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+    repo = client.get_repo(f"{args.owner}/{args.repo}")
+    # Get current repository settings.
+    current_settings = {
+        "branch_protection": {},
```
Why empty dict?

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+        )
+    else:
+        _LOG.warning("Running in non-interactive mode, skipping confirmation")
+    # Switch to default settings if requested.
+    settings.switch_default_branch_protection(repo, execute)
+    settings.switch_default_repository_settings(repo, execute)
```
How is this requested? Through which arg?

heanhsok commented on dev_scripts_helpers/github/settings/repo_settings.yaml:
```diff
@@ -0,0 +1,49 @@
+# Repository settings
+# - Basic settings: name, description, visibility, archive status
+# - Feature settings: issues, projects, wiki
+# - Merge settings: merge strategies and branch cleanup
+# - Security settings: automated fixes and vulnerability alerts
+# - Topics: repository categorization tags
```
Can we add the headings to the actual setting below as well? That way, it's easier to see the grouping

sandeepthalapanane commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+#!/usr/bin/env python3
+"""
+The script is designed to synchronize GitHub repository settings and branch
+protection rules from a settings manifest file. It requires certain
+dependencies to be present (e.g., `pygithub`) and thus it is executed as a
+dockerized executable.
```
I verified it with my own repository, but I couldn't test the organizational rules. I confirmed with the source code that these functions exist correctly. I would like to test it on an organizational test repo, if possible.

heanhsok replied:
We don't have org test repo but I think you can create one with personal account and test it

sandeepthalapanane commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,424 @@
+        self, repo: gh_repo.Repository, execute: bool = True
+    ) -> None:
+        """
+        Switch to default repository settings that exist in the repo but not in
+        settings manifest file.
```
The default settings refer to the settings present when creating a new repository.

heanhsok replied:
Hmm, so GitHub default setting? Can we use a YAML file to represent this rather than embedding the logic in the code? If we load that default setting from YAML, we can probably even use the same flow that was used to update the custom setting, so we don't need to rewrite separate logic.

sandeepthalapanane replied:
I could do that. There are default settings for repo, but not for branch protection. When we apply the switch default, do you want me to make a recommended set of settings, such as default branch protection settings, or completely skip applying default branch protection settings?

Example recommended branch protection settings:
```yaml
branch_protection:
  main:
    enforce_admins: true
    allow_force_pushes: false
    allow_deletions: false
```

sandeepthalapanane commented:
@heanhsok I tested the script with a test repository, and it works. I was unable to test the users and teams, but the backup file and switching to the default are working correctly.

heanhsok commented:
Could you help check this? @sandeepthalapanane 

```bash
user_1042@18f837e9d157:/app$ dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py --token_env_var GITHUB_PAT --input_file dev_scripts_helpers/github/settings/repo_settings.yaml --dry_run --owner HeaSoLab --repo demo-repository --backup
18:19:37 - INFO  hdbg.py init_logger:1013                               Saving log to file '/app/dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py.log'
18:19:37 - INFO  hdbg.py init_logger:1018                               > cmd='dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py --token_env_var GITHUB_PAT --input_file dev_scripts_helpers/github/settings/repo_settings.yaml --dry_run --owner HeaSoLab --repo demo-repository --backup'
18:19:38 - INFO  dockerized_sync_gh_repo_settings.py _main:527          Settings backed up to /app/tmp.settings.HeaSoLab.demo-repository.yaml
Are you sure you want to synchronize repository settings? [y/n] y
18:19:40 - INFO  dockerized_sync_gh_repo_settings.py apply_branch_protection:321 Would apply branch protection rules to main:
  strict: True
  contexts: ['ci/lint', 'ci/fasttests']
  enforce_admins: True
  dismiss_stale_reviews: True
  require_code_owner_reviews: True
  required_approving_review_count: 1
  user_push_restrictions: ['admin1', 'admin2']
  team_push_restrictions: ['maintainers', 'developers']
18:19:40 - INFO  dockerized_sync_gh_repo_settings.py apply_repo_settings:415 Would apply repository settings:
  name: test-repo
  description: Test repository with branch protection and settings
  homepage: https://github.com/test-repo
  private: True
  has_issues: True
  has_projects: True
  has_wiki: False
  allow_squash_merge: True
  allow_merge_commit: False
  allow_rebase_merge: True
  archived: False
Traceback (most recent call last):
  File "/app/dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py", line 556, in <module>
    _main(_parse())
  File "/app/dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py", line 551, in _main
    settings.apply_repo_settings(repo, args.dry_run)
  File "/app/dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py", line 419, in apply_repo_settings
    if enable_security_fixes is not None:
       ^^^^^^^^^^^^^^^^^^^^^
UnboundLocalError: cannot access local variable 'enable_security_fixes' where it is not associated with a value
```

heanhsok commented on dev_scripts_helpers/github/settings/default_settings.yaml:
```diff
@@ -0,0 +1,15 @@
+# Default settings for a repository.
```
We can this `default_repo_settings.yaml`

heanhsok commented on dev_scripts_helpers/github/settings/repo_settings.yaml:
```diff
@@ -0,0 +1,59 @@
+# Repository settings
```
1) I'm thinking we can have a file named template_repo_settings.yaml, which includes all the available setting options and serves as a reference.

2) If we want to customize the settings for a specific repository, like //helpers, we can duplicate the template file, rename it to something like repo_settings.HeaSoLab.demo-repository.yaml, customize it, and use that file to update the repository's settings.

3) Additionally, we can have another file that contains only the common settings shared across all repositories (e.g., merge rules and branch protection). This file should be repo-agnostic, meaning it doesn't include any repository-specific details (such as name, description, users, ...etc) and can be safely applied to any repo. However, settings not specified in this common file should remain unchanged and should not be overridden. 

Mostly, we'll use 3 to sync the settings.

WDYT?

sandeepthalapanane replied:
We already have the first configuration file. The second one is user-specific and can be customized based on the `repo_settings.yaml`, the user will create it. We also have default settings in place. Additionally, I've created a third YAML file that contains common settings. As I mentioned earlier, I've implemented `github.NotSet` to ensure that we do not alter any settings that are not specified in the YAML file.

heanhsok commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -0,0 +1,557 @@
+    else:
+        _LOG.warning("Running in non-interactive mode, skipping confirmation")
+    # Switch to default settings if requested.
+    if args.switch_default:
```
I meant using command instead of flag. You can refer to this example

https://github.com/causify-ai/helpers/blob/d228dfd26f558b40829fdc9e4bf7540737c0c078/main_pytest.py#L21-L26

sandeepthalapanane commented:
@heanhsok 

I have completely restructured the code following the review by @gpsaggese. 

**The new architecture flow includes:**

1. Two distinct sub-commands: **export** and **sync**.

2. Export Pipeline: get settings → validate →save.

3. Sync Pipeline: load → validate → diff → log (dry-run) / backup → apply.

4. The central logic has been encapsulated in a private class called `_RepoAndBranchSettings`, which builds repository and branch configurations once and reuses them.

5. Backups are now created unconditionally in the current working directory, ensuring that restore points are always available.

6. The `switch_default` has been removed. It seems to be confusing and likely unnecessary, and we already have a common settings file, which looks enough.

**Improvements made:**

1. There is now comprehensive logging for dry-run operations, allowing users to see exactly what would change. Additionally, various logs are available for debug mode to better understand the flow of the code.

2. I have introduced key validation for YAML files (_validate_settings) to catch any unknown setting parameters.

3. Helper functions (_build_settings, _log_settings) have been factored out to eliminate duplicate code.@heanhsok I have completely restructured the code after @gpsaggese's review. 

4. I have also added `current` and `target` prefixes to variable names for clarity. I made sure to use variable names like `repo_and_branch_settings`, `repo_settings`, and `branch_settings` to avoid confusion.  For example, `repo_and_branch_settings` contains both repository and branch protection setting dictionaries, and `repo_settings` contains only repository settings. 

5. I have added more detailed comments, doc strings, and examples.

gpsaggese commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -59,9 +107,15 @@ def get_repository_settings(
         :param repo: GitHub repository object
-        :return: dictionary containing repository settings
+        :return: dictionary containing current settings of the
+            repository E.g., { "name": str, "default_branch": str,
```
The linter probably messed this up.

Can we make a reference to the data structure REPO_SETTING_KEYS in the comment for DRY?

gpsaggese commented on dev_scripts_helpers/github/dockerized_sync_gh_repo_settings.py:
```diff
@@ -88,17 +142,44 @@ def get_branch_protection_settings(
         :param repo: GitHub repository object
         :return: dictionary containing branch protection settings
+        E.g.,
+        {
```
Let's add a ``` to make it markdown

But can we make reference to data structures?

gpsaggese reviewed (commented):
Looks good. Do we have a doc about the high level design of this?
